homeassistant/helpers/entity.py

```diff
@@ -842,9 +855,45 @@ def _async_write_ha_state(self) -> None:
             return
 
         start = timer()
-        state, attr = self._async_generate_attributes()
+        state, capabilities, shadowed_attr, attr = self._async_generate_attributes()
         end = timer()
 
+        if entry:
+            # Make sure capabilities in the entity registry are up to date. Capabilities
+            # include capability attributes, device class and supported features
+            original_device_class: str | None = shadowed_attr[ATTR_DEVICE_CLASS]
+            supported_features: int = attr.get(ATTR_SUPPORTED_FEATURES) or 0
+            if (
+                capabilities != entry.capabilities
+                or original_device_class != entry.original_device_class
+                or supported_features != entry.supported_features
+            ):
+                if not self.__capabilities_updated_at_reported:
+                    time_now = hass.loop.time()
+                    capabilities_updated_at = self.__capabilities_updated_at
+                    capabilities_updated_at.append(time_now)
+                    while time_now - capabilities_updated_at[0] > 3600:
+                        capabilities_updated_at.popleft()
+                    if len(capabilities_updated_at) >= CAPABILITIES_UPDATE_LIMIT:
```

Comment:
The deque can't be longer than the limit since we set a max length to the deque of the same limit.

Reply:
Right, so do you suggest to check for equality instead?